posts/inside-rust/2024-12-13-this-development-cycle-in-cargo-1.84.md

```diff
@@ -0,0 +1,400 @@
+We decided we should keep the `bench` profile matching `release` though we recognize there is room to explore improving user workflows for profiling.
+
+[foxtran](https://github.com/foxtran) restarted the conversation on changing the defaults for `release` to improve runtime performance in
+[#11298](https://github.com/rust-lang/cargo/issues/11298).
+
+Potential changes include
+- Enabling LTO, whether thin or fat
+- Reducing the codegen-units
+- Increasing the opt-level
+
+While `release` builds aren't focused on fast compile-times,
+there is still a point of diminishing returns in trading off compile-time for runtime performance.
+While `release` is generally geared towards production builds,
+there are cases where `dev` is too slow for development.
+[weihanglo](https://github.com/weihanglo) ran the numbers of LTO and codegen-units for Cargo in
+[#14719](https://github.com/rust-lang/cargo/issues/14719).
+From those numbers, it seems like thin LTO is an easy win.
+
+One option is for a `release-fast` or `release-heavy` to be made.
+Adding new profiles may be a breaking change though and we'd have to carefully approach doing so.
+We also already have discoverability problems with `release` and it has a dedicated flag (`--release`).
+Without some kind of built-in integration, it seems like these policies would be best left for users.
+
+Whatever profile is used, one problem with LTO is that there are miscompilations which might prevent it from being a safe default (e.g. [#115344](https://github.com/rust-lang/rust/issues/115344)).
+
+<!-- 2024-11-05 -->
+On the other end of the spectrum is the `dev` profile.
+This profile serves two roles
+- Fast iteration time
+- Running code through a debugger
+
+It turns out that these can be at odds with each other.
+When running through a debugger, you often want the binary to behave like the source code and optimizations can get in the way.
+However, optimizations can reduce the amount of IR being processed, speeding up codegen.
+They can then also speed up proc macros, build scripts, and test runs.
+Maybe we can even design a optimization level focused on improving compile times at the cost of the debugger experience.
+Similarly, how much debug information you carry around in your binary can affect your build times.
+
+Looking at the [Rust 2023 survey results](https://blog.rust-lang.org/2024/02/19/2023-Rust-Annual-Survey-2023-results.html#challenges),
+improving compilation times and the debugging experience is neck and neck.
+The question is which debugging experience are they referring to?
+Those on the call mostly used "printf"-style debugging and would get benefit out of improving compilation time.
+Even if we surveyed people and found this representative of the Rust community (which [davidlattimore](https://github.com/davidlattimore/) did for a subset of the community on [Mastadon](https://mas.to/@davidlattimore/113484821980790635)),
```

Comment:
```suggestion
Even if we surveyed people and found this representative of the Rust community (which [davidlattimore](https://github.com/davidlattimore/) did for a subset of the community on [Fediverse](https://mas.to/@davidlattimore/113484821980790635)),
```

(Mastodon was misspelled, but also, using the more generic term that includes other implementations.)